### PR TITLE
Reduce deferred renderer layers size

### PR DESCRIPTION
**Pull request by Sorien**
- What does the pull request do?

reduces deferred renderer layers size from scene size to layer root size 

Checklist:

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

File: src/Avalonia.Visuals/Platform/IDrawingContextImpl.cs [hunk 1]
```diff
@@ -17,6 +17,11 @@ public interface IDrawingContextImpl : IDisposable
         /// </summary>
         Matrix Transform { get; set; }
 
+        /// <summary>
+        /// Gets or sets the base transform of the drawing context.
+        /// </summary>
+        Matrix BaseTransform { get; set; }
```
Comment (grokys) [inline src/Avalonia.Visuals/Platform/IDrawingContextImpl.cs]: What does "base transform" mean? Why is it different from Transform?

**Review by grokys — CHANGES_REQUESTED** [review]
Could you give a little more details about what this does and how it does it? I'm having difficulty understanding.

From what I understand this makes it so that layers will only be created as large as they need to be, rather than at the scene size as they previously were? That's definitely an improvement.

However, something like this needs unit tests. Particularly, I'm concerned what will happen if a layer needs to change size? Will the whole layer be redrawn? If not, what if the layer needs to expand to the left or the top? Will the existing layer content be shifted over correctly?

**Comment by Sorien** [comment]
yes it creates layers as large as needs to be there is no change in drawings, layer resizing and so on, BaseTransform is Translation Matrix that translate all drawings operation to fit layer

original 
layer 1200x800, visual bounds [100, 100, 200, 200] visual is drawn at [100, 100, 200, 200] in layer, during layer combine layer is drawn at 0, 0

new 
layer 200x200, visual bounds [100, 100, 200, 200], (BaseTransform - translation matrix [1, 0 , -100; 0, 1, -100; 0, 0, 1]), visual is drawn at [0, 0, 200, 200] in layer, during layer combine layer is drawn at 100, 100

i'm not sure how to write tests for this

File: src/Avalonia.Visuals/Platform/IDrawingContextImpl.cs (hunk 1 above)
Comment (Sorien) [reply src/Avalonia.Visuals/Platform/IDrawingContextImpl.cs]: its just another tranformation matrix (pre transform/pre translation), that translate all drawings to new position, i didn't want to change all places where Transform is set

**Comment by Sorien** [comment]
I've tried to create way more complex sample with like hundreds of layers, and you have right it **creates artefacts when layers are resizing/moving** and it's a difficult to debug couse i'm getting 

```
  | Name | Value | Type
-- | -- | -- | --
▶ | $exception | {System.NullReferenceException: Object reference not set to an instance of an object.    at Avalonia.Shared.PlatformSupport.StandardRuntimePlatform.UnmanagedBlob.DoDispose() in D:\private\Avalonia\src\Shared\PlatformSupport\StandardRuntimePlatform.cs:line 83    at Avalonia.Shared.PlatformSupport.StandardRuntimePlatform.UnmanagedBlob.Finalize() in D:\private\Avalonia\src\Shared\PlatformSupport\StandardRuntimePlatform.cs:line 111} | System.NullReferenceException
```
all the time so there is probably something else wrong with that implementation?!,

just use it as reference for possible optimization as far as it will take way more work to do it right

**Pull request closed without merging**